src/connectionbenchmark/benchmarktaskdownload.cpp

```diff
@@ -0,0 +1,67 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "benchmarktaskdownload.h"
+#include "constants.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "networkrequest.h"
+
+#include <QByteArray>
+
+namespace {
+Logger logger(LOG_MAIN, "BenchmarkTaskDownload");
+}
+
+BenchmarkTaskDownload::BenchmarkTaskDownload(const QString& fileUrl)
+    : BenchmarkTask(Constants::BENCHMARK_MAX_DURATION_DOWNLOAD),
+      m_fileUrl(fileUrl) {
+  MVPN_COUNT_CTOR(BenchmarkTaskDownload);
+}
+
+BenchmarkTaskDownload::~BenchmarkTaskDownload() {
+  MVPN_COUNT_DTOR(BenchmarkTaskDownload);
+}
+
+void BenchmarkTaskDownload::runInternal() {
+  logger.debug() << "Run download benchmark";
+
+  connect(this, &BenchmarkTask::stateChanged, this,
+          &BenchmarkTaskDownload::handleState);
+}
+
+void BenchmarkTaskDownload::handleState(BenchmarkTask::State state) {
+  logger.debug() << "Handle state" << state;
+
+  if (state == BenchmarkTask::StateActive) {
+    m_request = NetworkRequest::createForGetUrl(this, m_fileUrl);
+    connect(m_request, &NetworkRequest::requestFailed, this,
+            &BenchmarkTaskDownload::downloadReady);
+    connect(m_request, &NetworkRequest::requestCompleted, this,
+            [&](const QByteArray& data) {
+              downloadReady(QNetworkReply::NoError, data);
+            });
+  } else if (state == BenchmarkTask::StateInactive && m_request) {
+    m_request->abort();
+    m_request = nullptr;
+  }
+}
+
+void BenchmarkTaskDownload::downloadReady(QNetworkReply::NetworkError error,
+                                          const QByteArray& data) {
+  logger.debug() << "Download ready" << error;
+
+  if (this->state() == BenchmarkTask::StateActive) {
+    this->stop();
```

Comment:
Why calling stop() This will trigger line 46 ad the request will be aborted.

Reply:
I guess you are doing it to set `m_executionTime`. If yes, you don't need it.
What about if you remove `m_executionTime` and you change `executionTime()` to return ` m_elapsedTimer.elapsed();` ?

Reply:
Yes, that was my intention — I changed it!